Template localization documentation (#4603)
dotnet/templating: docs/Localization.md

# Template localization 

> The templates are localizable. If the template is localized for the language matching the current locale, its elements will appear in that language in UX including `dotnet new` and Visual Studio.

Reviewer: ```suggestion
.NET Templates are localizable. If a template is localized for the language matching the current locale, its elements will appear in that language in hosts that use the Template Engine libraries, including `dotnet new` and the Visual Studio New Project Dialog.
```

Localizable elements are:
- name 
- author
- description
- symbols
  - description
  - display name
  - description and display name for choices for  choice parameters
- post actions
  - description
  - manual instructions

> Localization files should be put to `localize` folder inside `.template.config`. Format of localization file is JSON, there is one file per language.

Reviewer: ```suggestion
Localization files should be located inside a `.template-config\localize` folder. The format of the localization file is JSON, and there should be one file per language.
```

> Naming convention for localization files is: `templatestrings.<lang code>.json`. Language code should match name of [cultures](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.name?view=net-6.0#system-globalization-cultureinfo-name).

Reviewer: ```suggestion
The naming convention for localization files is: `templatestrings.<lang code>.json`, where the `lang code` should match one of the CultureInfo [names](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.name?view=net-6.0#system-globalization-cultureinfo-name) from the list provided by [GetCultures](https://docs.microsoft.com/dotnet/api/system.globalization.cultureinfo.getcultures?view=net-6.0).
```

> Content of JSON are key-value pairs, where:

Reviewer: ```suggestion
The structure of the localization JSON files are key-value pairs, where:
```

> - key is the reference to element of `template.json` to be localized. if the element is child, the full path using `/` delimiter should be given, example: `symbols/Framework/choices/netstandard2.1/description`.

Reviewer: ```suggestion
- The key is the reference to an element of `template.json` to be localized. If the element is child, the full path using `/` delimiter should be given, example: `symbols/Framework/choices/netstandard2.1/description`.
```

side note - was there a design on these paths? I wonder if [json path](https://github.com/json-path/JsonPath) was evaluated as an opion

> - value is localization of the element given in the key. 

Reviewer: ```suggestion
- The value is the localization of the element given in the key. 
```

Examples of localization files can be found [here](https://github.com/dotnet/templating/tree/main/template_feed/Microsoft.DotNet.Common.ProjectTemplates.7.0/content/ClassLibrary-CSharp/.template.config/localize).

> These files can be parsed by template engine when loading information about the template. Template engine API returns the information for localizable properties in language matching the current UI culture (if localization is available).

Reviewer: ```suggestion
These files are be parsed by the template engine when loading information about the template. The Template Engine API returns information for localizable properties in language matching the current UI culture (if localization is available) transparently, without explicit user action.
```

### Post action localization

> Ensure to add `id` property for all post actions in `template.json` prior setting up the localization. `id` should be unique within the template. 

Reviewer: ```suggestion
Ensure that all post actions in `template.json` have an `id` property prior setting up the localization. The `id` should be unique within the template. 
```

It might be a nice thing to add to validate this in the MSBuild Tasks. I might look at adding this :)

Author: It is already validated in MSBuild task.

Reviewer: Ah, I see - it happens via the ` LocalizationKeyIsNotUniqueException` during the template string extraction process.

Without the `id` property the localization files cannot be created.
In case the post action uses more than one `manualInstructions`, `id` should be also added for each manual instruction.

## Automatic generation of localization files

> The localization files can be generated automatically on the build using `LocalizeTemplates` MSBuild task from `Microsoft.TemplateEngine.Tasks` package.

Reviewer: ```suggestion
The localization files can be generated automatically as part of a build using the `LocalizeTemplates` MSBuild task from the [`Microsoft.TemplateEngine.Tasks`](https://www.nuget.org/packages/Microsoft.TemplateEngine.Tasks) package.
```

> The task is meant to be used in template package project and will create the JSON files after the project is build.

Reviewer: ```suggestion
The task is meant to be used in template package project and will create the JSON files after the project is built.
```

The task supports the configuration using following properties:
- `LocalizableTemplatesPath` - the folder containing templates to be localized. Default value: `.`.
- `TemplateLanguages` - the languages to create files for. Default value: `cs;de;en;es;fr;it;ja;ko;pl;pt-BR;ru;tr;zh-Hans;zh-Hant`.

Example of usage:

> ```xml
>  <ItemGroup>
>     <PackageReference Include="Microsoft.TemplateEngine.Tasks" Version="1.0.0.0" PrivateAssets="all" IsImplicitlyDefined="true" />
>  </ItemGroup>
>  <LocalizeTemplates>true</LocalizeTemplates>
>  <TemplateLanguages>en;de</TemplateLanguages>

Reviewer: ```suggestion
```xml
<PropertyGroup>
   <LocalizeTemplates>true</LocalizeTemplates>
   <TemplateLanguages>en;de</TemplateLanguages>
</PropertyGroup>

<ItemGroup>
   <PackageReference Include="Microsoft.TemplateEngine.Tasks" Version="1.0.0.0" PrivateAssets="all" IsImplicitlyDefined="true" />
</ItemGroup>
```

  - manual instructions

Localization files should be located inside a `.template-config\localize` folder. The format of the localization file is JSON, and there should be one file per language.
The naming convention for localization files is: `templatestrings.<lang code>.json`, where the `lang code` should match one of the CultureInfo [names](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.name?view=net-6.0#system-globalization-cultureinfo-name) from the list provided by [GetCultures](https://docs.microsoft.com/dotnet/api/system.globalization.cultureinfo.getcultures?view=net-6.0).
The structure of the localization JSON files are key-value pairs, where:
- The key is the reference to an element of `template.json` to be localized. If the element is child, the full path using `/` delimiter should be given, example: `symbols/Framework/choices/netstandard2.1/description`.
- The value is the localization of the element given in the key. 

Examples of localization files can be found [here](https://github.com/dotnet/templating/tree/main/template_feed/Microsoft.DotNet.Common.ProjectTemplates.7.0/content/ClassLibrary-CSharp/.template.config/localize).

> These files are be parsed by the template engine when loading information about the template. The Template Engine API returns information for localizable properties in language matching the current UI culture (if localization is available) transparently, without explicit user action.

Reviewer 2: Is there a grammatical mistake here? I could be reading this wrong as well.
```suggestion
These files are to be parsed by the template engine when loading information about the template. The Template Engine API returns information for localizable properties in language matching the current UI culture (if localization is available) transparently, without explicit user action.
```

The task is meant to be used in template package project and will create the JSON files after the project is built.
The task supports the configuration using following properties:
- `LocalizableTemplatesPath` - the folder containing templates to be localized. Default value: `.`.
- `TemplateLanguages` - the languages to create files for. Default value: `cs;de;en;es;fr;it;ja;ko;pl;pt-BR;ru;tr;zh-Hans;zh-Hant`.

Example of usage:
```xml
<PropertyGroup>
   <LocalizeTemplates>true</LocalizeTemplates>
   <TemplateLanguages>en;de</TemplateLanguages>
</PropertyGroup>

<ItemGroup>
   <PackageReference Include="Microsoft.TemplateEngine.Tasks" Version="1.0.0.0" PrivateAssets="all" IsImplicitlyDefined="true" />
</ItemGroup>
```

This example will generate localization files for English and German languages for all templates in the package.

## 1st party templates localization

Prerequisite: the repo should be onboarded for OneLocBuild to localize the template content.
See the [link](https://aka.ms/AllAboutLoc) for more details.

### Repo is using arcade

`Microsoft.TemplateEngine.Tasks` is included to `Microsoft.DotNet.Arcade.Sdk`.

> To start localizing the templates, enable the following property in template package project: <UsingToolTemplateLocalizer>true</UsingToolTemplateLocalizer>. This enables generating localization JSON files on the build.

Reviewer 2: The property is not visible in markdown renderer. Hopefully this will fix it:
```suggestion
To start localizing the templates, enable the following property in template package project: `<UsingToolTemplateLocalizer>true</UsingToolTemplateLocalizer>`. This enables generating localization JSON files on the build.
```

### Repo is not using arcade
`Microsoft.TemplateEngine.Tasks` may be still manually added to engineering or directly to template package project as explained above.
In addition to that, localization handover process should be changed as following:
- `LocalizeTemplates` task generates ready files to be used including `<lang code>` in filename. OneLocBuild should receive the file without language code: `templatestrings.json`. The localization pipeline should rename the `templatestrings.en.json` to `templatestrings.json` and include it to localization project JSON to be translated.
- please work with your localization developer advocate to ensure that
  - localization files are handed back in JSON format and to original location
After the translation is done, the repo should receive the PR replacing original English values in generated JSON files with localized ones. It is important to ensure that files are handed over to original location (`.template.config\localize` folder).

It is recommended to setup a way to update `Microsoft.TemplateEngine.Tasks` package version, however it is not recommended to update it on each daily build. 

> The package is released with any build of dotnet/templating repo, though it might have no changes, so updating version on each build might be superfluous. It is recommended to update the version when the next preview or stable version is released.

Reviewer 2: ```suggestion
The package is released with each build of dotnet/templating repo, even if there are no changes. Therefore, updating the version on each build might be superfluous. It is recommended to update the version when the next preview or stable version is released.
```
